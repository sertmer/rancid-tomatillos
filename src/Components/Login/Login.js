import React, { Component } from 'react';
import './Login.scss';

class Login extends Component {
  constructor() {
    super()

    this.state = {
      name: '',
      password: ''
    }
  }

  handleChange = (e) => {
    this.setState({ [e.target.name]: e.target.value })
  }

  render() {
    return (
      <form>
        <label for='user-name'>Name</label>
        <input 
          className='login-input'
          value={this.state.name} 
          name='name'
          type='text'
          onChange={(e) => this.handleChange(e)}
        />
        <label for='user-password'>Password</label>
        <input 
          className='login-input'
          value={this.state.password} 
          name='password'
          type='password'
          onChange={(e) => this.handleChange(e)}
        />
        <button type='button'>Login</button>
      </form>
    )
  }
}

export default Login;