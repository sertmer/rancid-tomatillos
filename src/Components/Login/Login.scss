@import '../../scss/variables';

.loader-wrapper {
  display: flex;
  width: 100vw;
  flex-direction: column;
  align-items: center;
  justify-content: center;
}

.loading-text {
  color: #fff;
  font-family: $accent-font;
}

.projector {
  height: 60vh;
  width: 63vw;
}
