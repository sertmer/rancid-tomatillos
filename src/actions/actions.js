export const getMovies = movies => ({
  type: 'GET_MOVIES',
  movies
});

export const currentUser = user => ({
  type: 'SET_CURRENT_USER',
  name: user.user.name,
  id: user.user.id,
  ratings: []
});

export const loggedIn = loggedIn => ({
  type: 'LOG_IN',
  loggedIn
});

export const getCurrentMovie = movie => ({
  type: 'GET_MOVIE',
  movie
});

export const getRatings = ratings => ({
  type: 'GET_RATINGS',
  ratings
});

