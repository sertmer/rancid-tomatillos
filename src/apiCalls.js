export const fetchMovies = () => {
  return fetch('https://rancid-tomatillos.herokuapp.com/api/v1/movies')
    .then(response => {
      if(!response.ok) {
        throw Error('Error fetching movies')
      }
      return response.json()
    })
}

export const getUser = (email, password) => {
  const options = {
    method: 'POST',
    body: JSON.stringify({
      email,
      password
    }),
    headers: {
      'Content-Type': 'application/json'
    }
  }

  return fetch('https://rancid-tomatillos.herokuapp.com/api/v1/login', options)
    .then(res => {
      if(!res.ok) {
        throw Error('Something is not right, try again later')
      }
      return res.json()
    })
}
