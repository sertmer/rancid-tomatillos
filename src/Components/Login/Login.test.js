import React from 'react';
import { Login, mapState, mapDispatch } from './Login';
import { shallow } from 'enzyme';
import { currentUser, loggedIn } from '../../actions/actions';
import { getUser, getUserRatings } from '../../apiCalls';
jest.mock('../../apiCalls')


describe('Login', () => {
  let wrapper, mockEvent, mockGetUser

  describe('Login Component', () => {
    beforeEach(() => {
      mockEvent = { target: {name: 'email', value:'abc123@aol.com'} }
      mockGetUser = jest.fn()
      wrapper = shallow(<Login />)
    });

    it('Should match the snapshot when there is no error in state', () => {
      expect(wrapper.debug()).toMatchSnapshot()
    });

    it('should match the snapshot when there is an error in state', () => {
      wrapper.setState({ error: 'YOU DID SOMETHING WRONG!!!!!' })

      expect(wrapper.debug()).toMatchSnapshot();
    })

    it('Should setState when handleChange is called', () => {
      wrapper.setState({ email: '', password:'', error: '' });
      wrapper.instance().handleChange(mockEvent);

      expect(wrapper.state()).toEqual({ email: 'abc123@aol.com', password:'', error: '' })
    });

    it('Should call handleChange when email input is changed', () => {
      wrapper.instance().handleChange = jest.fn();
      const mockEmailEvent = { target: {email: 'email', value:'abc123@aol.com'} }

      wrapper.find('.login-input-email').simulate('change', mockEmailEvent);

      expect(wrapper.instance().handleChange).toHaveBeenCalledWith(mockEmailEvent)
    });

    it('Should call handleChange when password input is changed', () => {
      wrapper.instance().handleChange = jest.fn();
      const mockPasswordEvent = { target: {password: 'password', value:'password123'} }

      wrapper.find('.login-input-password').simulate('change', mockPasswordEvent);

      expect(wrapper.instance().handleChange).toHaveBeenCalledWith(mockPasswordEvent)
    });

    it('Should call getUser with an email and password when handleLogin is called', () => {
      const mockEvent = { preventDefault: jest.fn() };
      getUser.mockImplementation(() => {
        return Promise.resolve()
      });

      wrapper.setState({ email: 'abc123@aol.com', password: 'password123', error: '' });
      wrapper.find('button').simulate('click', mockEvent);

      expect(getUser).toHaveBeenCalledWith('abc123@aol.com', 'password123')
    });

    

    it('Should call getUserRatings with a userId when handleGetUserRatings is called', () => {
      getUserRatings.mockImplementation(() => {
        return Promise.resolve() 
      });
      const expected = 1;

      wrapper.instance().handleGetUserRatings(expected);

      expect(getUserRatings).toHaveBeenCalledWith(expected)
    });
  });

  describe('mapState', () => {

    it('Should return a login value of true', () => {
      const mockState = {
        loggedIn: true,
        movies: [ {movie: 'movie1'} ]
      };
      const expected = { isLoggedIn: true };

      const mappedProps = mapState(mockState);

      expect(mappedProps).toEqual(expected);
    })
  });

  describe('mapDispatch', () => {

    it('Should call mapDispatch with a user when handleLogin is called', () => {
      const mockDispatch = jest.fn();
      const actionToDispatch = currentUser({ user: { id: 8, name: 'Rick' } });

      const mappedProps = mapDispatch(mockDispatch);
      mappedProps.currentUser({ user: { id: 8, name: 'Rick' } });

      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });

    it('Should call mapDispatch with a loggedIn boolean when handleLogin is called', () => {
      const mockDispatch = jest.fn();
      const actionToDispatch = loggedIn(true);

      const mappedProps = mapDispatch(mockDispatch);
      mappedProps.loggedIn(true);

      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });
  });

})
