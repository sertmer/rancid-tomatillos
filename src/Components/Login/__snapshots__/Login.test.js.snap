// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login login Component Should match the snapshot 1`] = `ShallowWrapper {}`;
